### ISSUE TITLE
llvm/clang switch statement -fno-jump-table add to build scripts

### ISSUE THREAD

**Issue by jrfastab** [cilium/cilium#1024]
add -fno-jump-table to bpf scripts

**Comment by jrfastab** [comment]
should we have a bpf lint? That would be cool :)

**Comment by aalemayhu** [comment]
Sounds good, maybe something like [Clang-Tidy][0] could be used.

[0]: http://clang.llvm.org/extra/clang-tidy/

**Comment by borkmann** [comment]
Here's what I got from Yonghong:

BPF backend should always generate if statements for "switch" statement.

However, if people compile like this at linux's samples/bpf ...
```
 clang ... -O2 -S -emit-llvm test.c
 llc -march=bpf -filetype=obj test.ll
```
... the jump-table issue may appear. The reason is first clang command essentially for x86_64 and it does generate jump-tables in readonly section.

**Comment by jrfastab** [comment]
The specific case of 'clang ...' followed by 'llc' possible issue is noted above. My original concern however was unfounded because cilium will use 'clang ... -target bpf ...' to generate eBPF programs. When this is done the switch visiting code `SelectionDAGBuilder::visitSwitch` in llvm code base at,
   ``` ./lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp```
will only create jump tables when the backend has support for either `ISD::BR_JT` or `ISD::BRIND` both of these instructions are "expanded" in the eBPF backend. As seen in
 ```   ./lib/Target/BPF/BPFISelLowering.cpp```
implemented here as follows,
 ``` setOperationAction(ISD::BR_JT, MVT::Other, Expand);```
  ```setOperationAction(ISD::BRIND, MVT::Other, Expand);```
So in short nothing to see here :) just being paranoid.
